src/python/nxt_python_asgi.c

```diff
@@ -199,10 +199,133 @@ nxt_python_asgi_init(nxt_unit_init_t *init, nxt_python_proto_t *proto)
+        { "call_soon",          &ctx_data->loop_call_soon },
+        { "run_until_complete", &ctx_data->loop_run_until_complete },
+        { "create_future",      &ctx_data->loop_create_future },
+    };
+
+    loop = NULL;
+
+    asyncio = PyImport_ImportModule("asyncio");
+    if (nxt_slow_path(asyncio == NULL)) {
+        nxt_unit_alert(NULL, "Python failed to import module 'asyncio'");
+        nxt_python_print_exception();
+        goto fail;
+    }
 
+    event_loop_func = "get_loop";
+    runner_class = "Runner";
+    runner_ref = PyDict_GetItemString(PyModule_GetDict(asyncio),
+                                      runner_class);
+    if (nxt_slow_path(runner_ref == NULL)) {
+        nxt_unit_alert(NULL,
+                       "Python failed to get '%s' from module 'asyncio'",
+                       runner_class);
+        goto fail;
+    }
+
+    runner = PyObject_CallObject(runner_ref, NULL);
+    if (nxt_slow_path(runner == NULL)) {
+        nxt_unit_alert(NULL, "Python failed to call 'asyncio.%s'",
+                       runner_class);
+        goto fail;
+    }
+    ctx_data->runner = runner;
+    loop_ref = PyObject_GetAttrString(runner, event_loop_func);
+    loop = PyObject_CallObject(loop_ref, NULL);
+        if (nxt_slow_path(runner == NULL)) {
+        nxt_unit_alert(NULL, "Python failed to call 'Runner.%s'",
+                       event_loop_func);
+        goto fail;
+    }
+    for (i = 0; i < nxt_nitems(handlers); i++) {
+        obj = PyObject_GetAttrString(loop, handlers[i].key);
+        if (nxt_slow_path(obj == NULL)) {
+            nxt_unit_alert(NULL, "Python failed to get 'loop.%s'",
+                                 handlers[i].key);
+            goto fail;
+        }
+
+        *handlers[i].handler = obj;
+
+        if (nxt_slow_path(PyCallable_Check(obj) == 0)) {
+            nxt_unit_alert(NULL, "'loop.%s' is not a callable object",
+                                 handlers[i].key);
+            goto fail;
+        }
+    }
+
+    obj = PyObject_CallObject(ctx_data->loop_create_future, NULL);
+    if (nxt_slow_path(obj == NULL)) {
+        nxt_unit_alert(NULL, "Python failed to create Future ");
+        nxt_python_print_exception();
+        goto fail;
+    }
+
+    ctx_data->quit_future = obj;
+
+    obj = PyObject_GetAttrString(ctx_data->quit_future, "set_result");
+    if (nxt_slow_path(obj == NULL)) {
+        nxt_unit_alert(NULL, "Python failed to get 'future.set_result'");
+        goto fail;
+    }
+
+    ctx_data->quit_future_set_result = obj;
+
+    if (nxt_slow_path(PyCallable_Check(obj) == 0)) {
+        nxt_unit_alert(NULL, "'future.set_result' is not a callable object");
+        goto fail;
+    }
+
+    Py_DECREF(loop);
+    Py_DECREF(asyncio);
+
+    *pdata = ctx_data;
+
+    return NXT_UNIT_OK;
+
+fail:
+
+    nxt_python_asgi_ctx_data_free(ctx_data);
+
+    Py_XDECREF(loop);
+    Py_XDECREF(asyncio);
+
+    return NXT_UNIT_ERROR;
+}
+
+#else
```

Comment:
This function can also be compacted to a single one with small preprocessor in it.